File: examples/docs-code-snippets/src/main/java/examples/java/application/AwaitEventWaitForNext.java
package examples.java.application;

import com.ing.baker.runtime.javadsl.Baker;
import com.ing.baker.runtime.javadsl.EventInstance;
import examples.java.events.OrderPlaced;

import java.time.Duration;

public class AwaitEventWaitForNext {

    private final Baker baker;

    public AwaitEventWaitForNext(Baker baker) {
        this.baker = baker;
    }

    public void example(String recipeInstanceId, OrderPlaced orderPlaced) {
        var eventInstance = EventInstance.from(orderPlaced);

        // Pattern: Start waiting BEFORE firing the sensory event
        // waitForNext = true checks for *new* occurrences of the event
        var awaitFuture = baker.awaitEvent(recipeInstanceId, "ExpectedEvent", Duration.ofSeconds(5), true);
        
        // Fire the event
        baker.fireSensoryEventAndAwaitReceived(recipeInstanceId, eventInstance).join();
        
        // Now wait for the event to occur
        // Because we started listening *before* firing, we are guaranteed not to miss it
        awaitFuture.join(); 
    }
}

File: examples/docs-code-snippets/src/main/kotlin/examples/kotlin/application/AwaitEventWaitForNext.kt
package examples.kotlin.application

import com.ing.baker.runtime.javadsl.EventInstance
import com.ing.baker.runtime.kotlindsl.Baker
import examples.kotlin.events.OrderPlaced
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlin.time.Duration.Companion.seconds

class AwaitEventWaitForNext(private val baker: Baker) {

    suspend fun example(recipeInstanceId: String, orderPlaced: OrderPlaced) = coroutineScope {
        val orderPlacedEvent = EventInstance.from(orderPlaced)
        
        // Pattern: Start waiting BEFORE firing the sensory event
        // waitForNext = true checks for *new* occurrences of the event
        // We use async to start the "await" concurrently 
        val awaitDeferred = async { 
            baker.awaitEvent(recipeInstanceId, "ExpectedEvent", timeout = 5.seconds, waitForNext = true) 
        }
        
        // Fire the event
        baker.fireSensoryEventAndAwaitReceived(recipeInstanceId, orderPlacedEvent)
        
        // Now wait for the event to occur
        awaitDeferred.await()
    }
}
